File: WordPress/Classes/ViewRelated/Me/MeViewController.m
// Me contents:
//
// + (No Title)
// | Account Settings
// | Help & Support
//
// + (No Title)
// | Log out

#import "MeViewController.h"
#import "SettingsViewController.h"
#import "SupportViewController.h"
#import "UIImageView+AFNetworkingExtra.h"
#import "MeHeaderView.h"
#import "ContextManager.h"
#import "AccountService.h"
#import "WPAccount.h"
#import "LoginViewController.h"
#import <WordPress-iOS-Shared/WPTableViewCell.h>
#import <WordPress-iOS-Shared/WPTableViewSectionHeaderFooterView.h>
#import "HelpshiftUtils.h"
#import "WordPress-Swift.h"


typedef NS_ENUM(NSInteger, MeSectionSections)
{
    MeSectionsAccount = 0,
    MeSectionsExtra,
    MeSectionsWpCom,
    MeSectionsCount
};

typedef NS_ENUM(NSInteger, MeSectionAccount) {
    MeSectionAccountSettings = 0,
    MeSectionAccountNotifications,
    MeSectionAccountCount
};

typedef NS_ENUM(NSInteger, MeSectionExtra) {
    MeSectionExtraHelp = 0,
    MeSectionExtraAbout,
    MeSectionExtraCount
};

typedef NS_ENUM(NSInteger, MeSectionWpCom) {
    MeSectionWpComAuthentication = 0,
    MeSectionWpComCount
};


static NSString *const WPMeRestorationID = @"WPMeRestorationID";
static NSString *const MVCCellReuseIdentifier = @"MVCCellReuseIdentifier";

@interface MeViewController () <UIViewControllerRestoration, UITableViewDataSource, UITableViewDelegate, UIActionSheetDelegate>

@property (nonatomic, strong) UITableView   *tableView;
@property (nonatomic, strong) MeHeaderView  *headerView;
@property (nonatomic, strong) NSDictionary  *rowCountMap;

@end

@implementation MeViewController

+ (UIViewController *)viewControllerWithRestorationIdentifierPath:(NSArray *)identifierComponents coder:(NSCoder *)coder
{
    return [self new];
}

#pragma mark - LifeCycle Methods

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.title = NSLocalizedString(@"Me", @"Me page title");
        self.restorationIdentifier = WPMeRestorationID;
        self.restorationClass = [self class];

        // we want to observe for the account change notification even if the view is not visible
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(defaultAccountDidChange:)
                                                     name:WPAccountDefaultWordPressComAccountChangedNotification
                                                   object:nil];

        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(helpshiftUnreadCountUpdated:)
                                                     name:HelpshiftUnreadCountUpdatedNotification
                                                   object:nil];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    [self buildTableView];
    [self refreshAccountUserDetails];
    [self refreshHeaderView];
    [self refreshTableView];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:NO animated:animated];

    [HelpshiftUtils refreshUnreadNotificationCount];
}

#pragma mark - View Construction / Configuration

- (void)buildTableView
{
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStyleGrouped];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    self.tableView.rowHeight = WPTableViewDefaultRowHeight;
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.tableView registerClass:[WPTableViewCell class] forCellReuseIdentifier:MVCCellReuseIdentifier];
    [self.view addSubview:self.tableView];

    self.headerView = [[MeHeaderView alloc] initWithFrame:CGRectMake(0.0, 0.0, CGRectGetWidth(self.view.bounds), MeHeaderViewHeight)];

    [WPStyleGuide configureColorsForView:self.view andTableView:self.tableView];

    [self setupAutolayoutConstraints];
}

- (void)setupAutolayoutConstraints
{
    NSMutableDictionary *views = [@{@"tableView": self.tableView} mutableCopy];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|[tableView]|"
                                                                      options:0
                                                                      metrics:nil
                                                                        views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[tableView]|"
                                                                      options:0
                                                                      metrics:nil
                                                                        views:views]];
}

- (void)refreshAccountUserDetails
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];
    WPAccount *defaultAccount = [accountService defaultWordPressComAccount];
    if (!defaultAccount) {
        return;
    }

    // We want to keep email and default blog information up to date, this is a reasonable best place to do it
    [accountService updateUserDetailsForAccount:defaultAccount success:^{
        [self refreshHeaderViewEmail];
    } failure:nil];
}

#pragma mark - Header methods

- (void)refreshHeaderView
{
    WPAccount *defaultAccount = [self defaultAccount];

    if (defaultAccount) {
        self.tableView.tableHeaderView = self.headerView;
        [self.headerView setDisplayName:defaultAccount.displayName];
        [self.headerView setUsername:defaultAccount.username];
        [self.headerView setGravatarEmail:defaultAccount.email];
    }
    else {
        self.tableView.tableHeaderView = nil;
    }
}

- (void)refreshHeaderViewEmail
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];
    WPAccount *defaultAccount = [accountService defaultWordPressComAccount];
    [self.headerView setGravatarEmail:defaultAccount.email];
}

- (void)refreshTableView
{
    // Let's hide the Notifications Row, whenever there is no default account
    NSInteger accountRowCount = MeSectionAccountCount;
    if (!self.defaultAccount) {
        --accountRowCount;
    }
    
    self.rowCountMap = @{
        @(MeSectionsAccount)    : @(accountRowCount),
        @(MeSectionsExtra)      : @(MeSectionExtraCount),
        @(MeSectionsWpCom)      : @(MeSectionWpComCount)
    };
    
    [self.tableView reloadData];
}

- (WPAccount *)defaultAccount
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    AccountService *accountService  = [[AccountService alloc] initWithManagedObjectContext:context];
    
    return accountService.defaultWordPressComAccount;
}

- (UILabel *)helpshiftBadgeLabel
{
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 50, 30)];
    label.layer.masksToBounds = YES;
    label.layer.cornerRadius = 15;
    label.textAlignment = NSTextAlignmentCenter;
    label.backgroundColor = [WPStyleGuide newKidOnTheBlockBlue];
    label.textColor = [UIColor whiteColor];
    return label;
}


#pragma mark - UITableViewDataSource methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return MeSectionsCount;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.rowCountMap[@(section)] intValue] ?: 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    WPTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:MVCCellReuseIdentifier];
    [WPStyleGuide configureTableViewCell:cell];

    if (indexPath.section == MeSectionsAccount) {
        switch (indexPath.row) {
            case MeSectionAccountSettings:
                cell.textLabel.text = NSLocalizedString(@"Account Settings", @"");
                cell.textLabel.textAlignment = NSTextAlignmentLeft;
                cell.accessibilityLabel = @"Account Settings";
                cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
                break;
            case MeSectionAccountNotifications:
                cell.textLabel.text = NSLocalizedString(@"Notification Settings", @"");
                cell.textLabel.textAlignment = NSTextAlignmentLeft;
                cell.accessibilityLabel = @"Notification Settings";
                cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
                break;
        }
    } else if (indexPath.section == MeSectionsExtra) {
        switch (indexPath.row) {
            case MeSectionExtraHelp:
                cell.textLabel.text = NSLocalizedString(@"Help & Support", @"");
                cell.textLabel.textAlignment = NSTextAlignmentLeft;
                cell.accessibilityLabel = @"Help & Support";

                NSInteger unreadNotificationCount = [HelpshiftUtils unreadNotificationCount];
                if ([HelpshiftUtils isHelpshiftEnabled] && unreadNotificationCount > 0) {
                    UILabel *label = [self helpshiftBadgeLabel];
                    label.text = [NSString stringWithFormat:@"%ld", unreadNotificationCount];
                    cell.accessoryView = label;
                    cell.accessoryType = UITableViewCellAccessoryNone;
                } else {
                    cell.accessoryView = nil;
                    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
                }
                break;
            case MeSectionExtraAbout:
                cell.textLabel.text = NSLocalizedString(@"About", @"");
                cell.textLabel.textAlignment = NSTextAlignmentLeft;
                cell.accessibilityLabel = @"About";
                cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
                break;
        }
    } else if (indexPath.section == MeSectionsWpCom) {
        if (indexPath.row == MeSectionWpComAuthentication) {
            NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
            AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];
            WPAccount *defaultAccount = [accountService defaultWordPressComAccount];
            
            cell.accessoryType = UITableViewCellAccessoryNone;

            if (defaultAccount) {
                NSString *signOutString = NSLocalizedString(@"Disconnect from WordPress.com",
                                                            @"Label for disconnecting from WordPress.com account");
                cell.textLabel.text = signOutString;
                cell.accessibilityIdentifier = signOutString;
                [WPStyleGuide configureTableViewDestructiveActionCell:cell];
            }
            else {
                NSString *signInString = NSLocalizedString(@"Connect to WordPress.com",
                                                           @"Label for connecting to WordPress.com account");
                cell.textLabel.text = signInString;
                cell.accessibilityIdentifier = signInString;
                [WPStyleGuide configureTableViewActionCell:cell];
                cell.textLabel.textAlignment = NSTextAlignmentCenter;
            }
        }
    }
    return cell;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{
    WPTableViewSectionHeaderFooterView *header = [[WPTableViewSectionHeaderFooterView alloc] initWithReuseIdentifier:nil style:WPTableViewSectionStyleHeader];
    header.title = [self titleForHeaderInSection:section];
    return header;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    NSString *title = [self titleForHeaderInSection:section];
    if (!title) {
        return CGFLOAT_MIN;
    }
    
    return [WPTableViewSectionHeaderFooterView heightForHeader:title width:CGRectGetWidth(self.view.bounds)];
}

- (NSString *)titleForHeaderInSection:(NSInteger)section
{
    if (section == MeSectionsWpCom) {
        return NSLocalizedString(@"WordPress.com Account", @"WordPress.com sign-in/sign-out section header title");
    }
    return nil;
}

#pragma mark - UITableViewDelegate methods

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];

    if (indexPath.section == MeSectionsAccount) {
        switch (indexPath.row) {
            case MeSectionAccountSettings:
                [self navigateToAccountSettings];
                break;
            case MeSectionAccountNotifications:
                [self navigateToNotificationSettings];
                break;
        }
    } else if (indexPath.section == MeSectionsExtra) {
        switch (indexPath.row) {
            case MeSectionExtraHelp:
                [self navigateToHelp];
                break;
            case MeSectionExtraAbout:
                [self navigateToAbout];
                break;
        }
        
    } else if (indexPath.section == MeSectionsWpCom) {
        if (indexPath.row == MeSectionWpComAuthentication) {
            NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
            AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];
            WPAccount *defaultAccount = [accountService defaultWordPressComAccount];

            if (defaultAccount) {
                // Present the Sign out ActionSheet
                NSString *signOutTitle = NSLocalizedString(@"Disconnecting your account will remove all of @%@’s WordPress.com data from this device.",
                                                           @"Label for disconnecting WordPress.com account. The %@ is a placeholder for the user's screen name.");
                signOutTitle = [NSString stringWithFormat:signOutTitle, [defaultAccount username]];
                UIActionSheet *actionSheet;
                actionSheet = [[UIActionSheet alloc] initWithTitle:signOutTitle
                                                          delegate:self
                                                 cancelButtonTitle:NSLocalizedString(@"Cancel", @"")
                                            destructiveButtonTitle:NSLocalizedString(@"Disconnect", @"Button for confirming disconnecting WordPress.com account")
                                                 otherButtonTitles:nil];
                actionSheet.actionSheetStyle = UIActionSheetStyleDefault;

                if (IS_IPAD) {
                    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
                    [actionSheet showFromRect:[cell bounds] inView:cell animated:YES];
                } else {
                    [actionSheet showInView:self.view];
                }
            } else {
                LoginViewController *loginViewController = [[LoginViewController alloc] init];
                loginViewController.onlyDotComAllowed = YES;
                loginViewController.cancellable = YES;
                loginViewController.dismissBlock = ^(BOOL cancelled){
                    [self dismissViewControllerAnimated:YES completion:nil];
                };

                UINavigationController *loginNavigationController = [[UINavigationController alloc] initWithRootViewController:loginViewController];
                [self presentViewController:loginNavigationController animated:YES completion:nil];
            }
        }
    }
}

#pragma mark - Actions

- (void)navigateToAccountSettings
{
    [WPAnalytics track:WPAnalyticsStatOpenedSettings];

    SettingsViewController *settingsViewController = [[SettingsViewController alloc] initWithStyle:UITableViewStyleGrouped];
    [self.navigationController pushViewController:settingsViewController animated:YES];
}

- (void)navigateToNotificationSettings
{
    NSString *nibName = [NotificationSettingsViewController classNameWithoutNamespaces];
    NotificationSettingsViewController *settingsViewController = [[NotificationSettingsViewController alloc] initWithNibName:nibName bundle:nil];
    [self.navigationController pushViewController:settingsViewController animated:YES];
}

- (void)navigateToHelp
{
    SupportViewController *supportViewController = [[SupportViewController alloc] init];
    [self.navigationController pushViewController:supportViewController animated:YES];
}

- (void)navigateToAbout
{
    NSString *nibName               = [AboutViewController classNameWithoutNamespaces];
    AboutViewController *aboutVC    = [[AboutViewController alloc] initWithNibName:nibName bundle:nil];
    
    [self.navigationController pushViewController:aboutVC animated:YES];
}


#pragma mark - Notifications

- (void)defaultAccountDidChange:(NSNotification *)notification
{
    [self refreshHeaderView];
    [self refreshTableView];
}


#pragma mark - Action Sheet Delegate Methods

- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if (buttonIndex == 0) {
        // Sign out asynchronously so the popover animation can finish on iPad #3667
        dispatch_async(dispatch_get_main_queue(), ^{
            // Sign out
            NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
            AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];

            [accountService removeDefaultWordPressComAccount];

            // reload all table view to update the header as well
            [self refreshTableView];
        });
    }
}

#pragma mark - Helpshift Notifications

- (void)helpshiftUnreadCountUpdated:(NSNotification *)notification
{
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:MeSectionExtraHelp inSection:MeSectionsExtra];
    [self.tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationNone];
}

@end

File: WordPress/Classes/ViewRelated/MeViewController.swift
import UIKit
import WordPressShared

class MeViewController: UITableViewController, UIViewControllerRestoration {
    static let restorationIdentifier = "WPMeRestorationID"
    var tableViewModel = ImmuTable(sections: []) {
        didSet {
            if isViewLoaded() {
                tableView.reloadData()
            }
        }
    }

    static func viewControllerWithRestorationIdentifierPath(identifierComponents: [AnyObject], coder: NSCoder) -> UIViewController? {
        return self.init()
    }

    // MARK: - Table View Controller

    override init(style: UITableViewStyle) {
        super.init(style: style)
        navigationItem.title = NSLocalizedString("Me", comment: "Me page title")
        restorationIdentifier = self.dynamicType.restorationIdentifier
        restorationClass = self.dynamicType
        clearsSelectionOnViewWillAppear = false
    }

    required convenience init() {
        self.init(style: .Grouped)
        let notificationCenter = NSNotificationCenter.defaultCenter()
        notificationCenter.addObserver(self, selector: "refreshModelWithNotification:", name: WPAccountDefaultWordPressComAccountChangedNotification, object: nil)
        notificationCenter.addObserver(self, selector: "refreshModelWithNotification:", name: WPAccountDefaultWordPressComAccountDetailsUpdatedNotification, object: nil)
        notificationCenter.addObserver(self, selector: "refreshModelWithNotification:", name: HelpshiftUnreadCountUpdatedNotification, object: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        ImmuTable.registerRows([
            NavigationItemRow.self,
            BadgeNavigationItemRow.self,
            ButtonRow.self,
            DestructiveButtonRow.self
            ], tableView: self.tableView)

        reloadViewModel()
        // FIXME: @koke 2015-12-17
        // See https://github.com/wordpress-mobile/WordPress-iOS/issues/4416
        // The view controller should observe changes to account details
        // regardless of who asked for them.
        // For now I'm just porting this to Swift as it is.
        refreshAccountDetails()

        WPStyleGuide.resetReadableMarginsForTableView(tableView)
        WPStyleGuide.configureColorsForView(view, andTableView: tableView)
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        HelpshiftUtils.refreshUnreadNotificationCount()
        animateDeselectionInteractively()
    }

    func reloadViewModel() {
        let account = defaultAccount()
        let loggedIn = account != nil
        let badgeCount = HelpshiftUtils.isHelpshiftEnabled() ? HelpshiftUtils.unreadNotificationCount() : 0

        // Warning: If you set the header view after the table model, the
        // table's top margin will be wrong.
        //
        // My guess is the table view adjusts the height of the first section
        // based on if there's a header or not.
        tableView.tableHeaderView = headerView(account)
        tableViewModel = tableViewModel(loggedIn, helpshiftBadgeCount: badgeCount)
    }

    func headerView(account: WPAccount?) -> MeHeaderView? {
        guard let account = account else { return nil }

        let header = cachedHeaderView
        header.setDisplayName(account.displayName)
        header.setUsername(account.username)
        header.setGravatarEmail(account.email)
        return header
    }

    func tableViewModel(loggedIn: Bool, helpshiftBadgeCount: Int) -> ImmuTable {
        let myProfile = NavigationItemRow(
            title: NSLocalizedString("My Profile", comment: "Link to My Profile section"),
            action: pushMyProfile())

        let accountSettings = NavigationItemRow(
            title: NSLocalizedString("Account Settings", comment: "Link to Account Settings section"),
            action: pushAccountSettings())

        let notificationSettings = NavigationItemRow(
            title: NSLocalizedString("Notification Settings", comment: "Link to Notification Settings section"),
            action: pushNotificationSettings())

        let helpAndSupport = BadgeNavigationItemRow(
            title: NSLocalizedString("Help & Support", comment: "Link to Help section"),
            badgeCount: helpshiftBadgeCount,
            action: pushHelp())

        let about = NavigationItemRow(
            title: NSLocalizedString("About", comment: "Link to About section (contains info about the app)"),
            action: pushAbout())

        let logIn = ButtonRow(
            title: NSLocalizedString("Connect to WordPress.com", comment: "Label for connecting to WordPress.com account"),
            action: presentLogin())

        let logOut = DestructiveButtonRow(
            title: NSLocalizedString("Disconnect from WordPress.com", comment: "Label for disconnecting from WordPress.com account"),
            action: confirmLogout())

        let wordPressComAccount = NSLocalizedString("WordPress.com Account", comment: "WordPress.com sign-in/sign-out section header title")

        if loggedIn {
            if Feature.enabled(.MyProfile) {
                return ImmuTable(
                    sections: [
                        ImmuTableSection(rows: [
                            myProfile,
                            accountSettings,
                            notificationSettings
                            ]),
                        ImmuTableSection(rows: [
                            helpAndSupport,
                            about
                            ]),
                        ImmuTableSection(
                            headerText: wordPressComAccount,
                            rows: [
                                logOut
                            ])
                    ])
            } else {
                return ImmuTable(
                    sections: [
                        ImmuTableSection(rows: [
                            accountSettings,
                            notificationSettings
                            ]),
                        ImmuTableSection(rows: [
                            helpAndSupport,
                            about
                            ]),
                        ImmuTableSection(
                            headerText: wordPressComAccount,
                            rows: [
                                logOut
                            ])
                    ])
            }
        } else { // Logged out
            return ImmuTable(
                sections: [
                    ImmuTableSection(rows: [
                        accountSettings,
                        ]),
                    ImmuTableSection(rows: [
                        helpAndSupport,
                        about
                        ]),
                    ImmuTableSection(
                        headerText: wordPressComAccount,
                        rows: [
                            logIn
                        ])
                ])
        }
    }
    // MARK: Table View Data Source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return tableViewModel.sections.count
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tableViewModel.sections[section].rows.count
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let row = tableViewModel.rowAtIndexPath(indexPath)
        let cell = tableView.dequeueReusableCellWithIdentifier(row.reusableIdentifier, forIndexPath: indexPath)

        row.configureCell(cell)

        return cell
    }

    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return tableViewModel.sections[section].headerText
    }

    // MARK: Table View Delegate

    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let row = tableViewModel.rowAtIndexPath(indexPath)
        row.action?(row)
    }

    override func tableView(tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let view = WPTableViewSectionHeaderFooterView(reuseIdentifier: nil)
        view.title = self.tableView(tableView, titleForHeaderInSection: section)
        return view
    }

    override func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        let title = self.tableView(tableView, titleForHeaderInSection: section)
        return WPTableViewSectionHeaderFooterView.heightForHeader(title, width: tableView.frame.width)
    }

    // MARK: - Actions

    func pushMyProfile() -> ImmuTableActionType {
        return { [unowned self] row in
            WPAppAnalytics.track(.OpenedMyProfile)
            let controller = MyProfileViewController()
            controller.account = self.defaultAccount()
            self.navigationController?.pushViewController(controller, animated: true)
        }
    }

    func pushAccountSettings() -> ImmuTableActionType {
        return { [unowned self] row in
            WPAppAnalytics.track(.OpenedAccountSettings)
            let controller = SettingsViewController()
            self.navigationController?.pushViewController(controller, animated: true)
        }
    }

    func pushNotificationSettings() -> ImmuTableActionType {
        return { [unowned self] row in
            let controller = NotificationSettingsViewController()
            self.navigationController?.pushViewController(controller, animated: true)
        }
    }

    func pushHelp() -> ImmuTableActionType {
        return { [unowned self] row in
            let controller = SupportViewController()
            self.navigationController?.pushViewController(controller, animated: true)
        }
    }

    func pushAbout() -> ImmuTableActionType {
        return { [unowned self] row in
            let controller = AboutViewController()
            self.navigationController?.pushViewController(controller, animated: true)
        }
    }

    func presentLogin() -> ImmuTableActionType {
        return { [unowned self] row in
            let controller = LoginViewController()
            controller.onlyDotComAllowed = true
            controller.cancellable = true
            controller.dismissBlock = { [unowned self] _ in
                self.dismissViewControllerAnimated(true, completion: nil)
            }

            let navigation = UINavigationController(rootViewController: controller)
            self.presentViewController(navigation, animated: true, completion: nil)
        }
    }

    func confirmLogout() -> ImmuTableActionType {
        return { [unowned self] row in
            let format = NSLocalizedString("Disconnecting your account will remove all of @%@’s WordPress.com data from this device.", comment: "Label for disconnecting WordPress.com account. The %@ is a placeholder for the user's screen name.")
            let title = String(format: format, self.defaultAccount()!.username)
            let alert = UIAlertController(title: title, message: nil, preferredStyle: .Alert)

            let cancel = UIAlertAction(
                title: NSLocalizedString("Cancel", comment: ""),
                style: .Cancel,
                handler: nil)
            let disconnect = UIAlertAction(
                title: NSLocalizedString("Disconnect", comment: "Button for confirming disconnecting WordPress.com account"),
                style: .Destructive,
                handler: { [unowned self] _ in
                self.logOut()
            })

            alert.addAction(cancel)
            alert.addAction(disconnect)

            self.presentViewController(alert, animated: true, completion: nil)
            self.tableView.deselectSelectedRowWithAnimation(true)
        }
    }

    // MARK: - Notification observers

    func refreshModelWithNotification(notification: NSNotification) {
        // Regenerate the header view if the account has changed
        cachedHeaderView = MeHeaderView()

        reloadViewModel()
    }

    // MARK: - Helpers

    // FIXME: (@koke 2015-12-17) Not cool. Let's stop passing managed objects
    // and initializing stuff with safer values like userID
    func defaultAccount() -> WPAccount? {
        let context = ContextManager.sharedInstance().mainContext
        let service = AccountService(managedObjectContext: context)
        let account = service.defaultWordPressComAccount()
        // Again, ! isn't cool, but let's keep it for now until we refactor the VC
        // initialization parameters.
        return account
    }

    func refreshAccountDetails() {
        guard let account = defaultAccount() else { return }
        let context = ContextManager.sharedInstance().mainContext
        let service = AccountService(managedObjectContext: context)
        service.updateUserDetailsForAccount(account, success: { [weak self] in
            self?.reloadViewModel()
            }, failure: { _ in })
    }

    func logOut() {
        let context = ContextManager.sharedInstance().mainContext
        let service = AccountService(managedObjectContext: context)
        service.removeDefaultWordPressComAccount()
    }

    lazy var cachedHeaderView = MeHeaderView()
}
